### PR TITLE
Add Ord implementation for KeepAlive

### PR DESCRIPTION
**Pull request by roignpar**
This makes it cleaner and more straightforward to pick the largest when multiple `KeepAlive` are combined.

### CODE REVIEW

File: core/src/protocols_handler/mod.rs [hunk 1]
```diff
@@ -434,3 +434,24 @@ impl KeepAlive {
...
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for KeepAlive {
+    fn cmp(&self, other: &KeepAlive) -> Ordering {
+        use self::KeepAlive::*;
+
+        match (self, other) {
+            (Now, Now) | (Forever, Forever) => Ordering::Equal,
```
Comment (tomaka) [inline core/src/protocols_handler/mod.rs]: Is infinite equal to infinite? :laughing:

**Comment by tomaka** [comment]
My comment was initially intended as a joke. However it's actually a non-trivial question.

The underlying problem is that the value of `Now` "changes automatically" over time.

Let me know what you think, but I think that `Now` should be treated as "minus infinity" instead, and we should only implement `PartialOrd` instead of `Ord`.

**Comment by roignpar** [comment]
I agree that we can consider `Now` to be "minus infinity" and assume that `Until` will always be greater.
In this case we can simply rearrange the variants of `KeepAlive` in order: `Now`, `Until`, `Forever` and derive `PartialOrd`. I can't think of a good reason to have an explicit `impl`. Is there one?

**Comment by tomaka** [comment]
The compiler can't do that automatically. The order of variants in an enum has absolutely no influence on the logic of any trait you derive.

**Comment by roignpar** [comment]
The `PartialOrd` docs say:
> When derived on enums, variants are ordered by their top-to-bottom declaration order.

https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html

**Comment by roignpar** [comment]
There could be a reason to `impl` instead of just deriving though: if a new variant is added to `KeepAlive` in the future in the wrong position, logic could possibly break in other places.

With an explicit `impl PartialOrd for KeepAlive`, the compiler will complain that the new variant is not handled. Simply deriving `PartialOrd` could cause the program to silently ignore bad results of comparing multiple `KeepAlive` when variants are not placed in the correct order.

Let me know what you think.

**Comment by tomaka** [comment]
I didn't know about that automatic derivation ordering thing. I would indeed prefer a manual implementation, in order to be explicit.

**Comment by tomaka** [comment]
Right now the `Ord` implementation isn't commutative. If `A` and `B` are both `Now`, then `A < B` and `B < A` will both be true.
I suggest we make `Now` equal to `Now` and `Forever` equal to `Forever`. However `Now` should always be inferior to `Until`, otherwise `A < B` could automatically become `B < A` over time, which we definitely don't want to happen.

**Comment by roignpar** [comment]
Is it worth adding some unit tests?

**Comment by tomaka** [comment]
I think it's fine. The code is straight-orward enough that there's no possible side-effect or corner case or logic issue that a unit test would cover.
